text/0074-stateful-multisig-pallet.md

```diff
@@ -0,0 +1,585 @@
+
+Let's start with a sequence diagram to illustrate the main operations of the Stateful Multisig.
+
+![multisig operations](https://github.com/asoliman92/RFCs/assets/2677789/4f2e8972-f3b8-4250-b75f-1e4788b35752)
+
+Notes on above diagram:
+
+* It's a 3 step process to execute a proposal. (Start Proposal --> Approvals --> Execute Proposal)
+* `Execute` is an explicit extrinsic for a simpler API. It can be optimized to be executed automatically after getting enough approvals.
+* Any user can create a multisig account and they don't need to be part of it. (Alice in the diagram)
+* A proposal is any extrinsic including control extrinsics (e.g. add/remove signer, change threshold,..etc).
+* Any multisig account signer can start a proposal on behalf of the multisig account. (Bob in the diagram)
+* Any multisig account owener can execute proposal if it's approved by enough signers. (Dave in the diagram)
+
+### State Transition Functions
+
+* `create_multisig` - Create a multisig account with a given threshold and initial signers. (Needs Deposit)
+
+```rust
+		/// Creates a new multisig account and attach signers with a threshold to it.
+		///
+		/// The dispatch origin for this call must be _Signed_. It is expected to be a nomral AccountId and not a
+		/// Multisig AccountId.
+		///
+		/// T::BaseCreationDeposit + T::PerSignerDeposit * signers.len() will be held from the caller's account.
+		///
+		/// # Arguments
+		///
+		/// - `signers`: Initial set of accounts to add to the multisig. These may be updated later via `add_signer`
+		/// and `remove_signer`.
+		/// - `threshold`: The threshold number of accounts required to approve an action. Must be greater than 0 and
+		/// less than or equal to the total number of signers.
+		///
+		/// # Errors
+		///
+		/// * `TooManySignatories` - The number of signatories exceeds the maximum allowed.
+		/// * `InvalidThreshold` - The threshold is greater than the total number of signers.
+		pub fn create_multisig(
+			origin: OriginFor<T>,
+			signers: BoundedBTreeSet<T::AccountId, T::MaxSignatories>,
+			threshold: u32,
+		) -> DispatchResult 
+```
+
+* `start_proposal` - Start a multisig proposal. (Needs Deposit)
+
+```rust
+		/// Starts a new proposal for a dispatchable call for a multisig account.
+		/// The caller must be one of the signers of the multisig account.
+		/// T::ProposalDeposit will be held from the caller's account.
+		///
+		/// # Arguments
+		///
+		/// * `multisig_account` - The multisig account ID.
+		/// * `call` - The dispatchable call to be executed.
+		///
+		/// # Errors
+		///
+		/// * `MultisigNotFound` - The multisig account does not exist.
+		/// * `UnAuthorizedSigner` - The caller is not an signer of the multisig account.
+		/// * `TooManySignatories` - The number of signatories exceeds the maximum allowed. (shouldn't really happen as it's the first approval)
+		pub fn start_proposal(
```

Comment:
A major hurdle for users, is that they need some offline communication to pass around the `call data`, so that they can verify that  the hash is actually matching, and for the last signer to execute the call.
 
While this is great for privacy, as in it's hard to guess as an external watcher what the call is about by knowing only the hash. Experience has shown that most users don't need this additional privacy, the reality is that they don't verify the `call data` and blindly approve multisig calls, trusting the original proposer. As a result, Multix and many other multisig management platforms use `multisig.asMulti` for the proposer call, hence exposing the call in the block and relying on an indexer to later decode the call, and show users what they are signing. 

If we add such a pallet to a system chain, I'd strongly advocate to include a mechanism to optionally store the call in the state, for `start_proposal` and then have it optional as well for the `execute_proposal` and retrieve it from the state is possible.

Reply:
I don't really have strong opinion here. I initially had it as a `RuntimeCall` and changed it in favor of the hash for both privacy and block footprint. I was thinking using preimage as well but maybe that's just complicating it.

> the reality is that they don't verify the call data and blindly approve multisig calls, trusting the original proposer.

Don't you think we can offload this part to the platforms instead to get the best of both worlds? Multix can get the call and hash it with the correct hash before calling the extrinsic. That will ensure seamless and private experience for correct users without sacrificing block footprint. It might not be the best for people who want to use it directly from polkadot.js for example but still good enough for most people who'll use some tool. What are your thoughts?

Reply:
We're building a decentralized network, yet for average users to be able to use multisigs we have to resort to centralized indexers (in the case of Multix) or IMHO worse, full blown web2 stacks, storing the callData. Letting such important part of the puzzle to centralized entities is certainly not the best of both worlds. If we let users choose, then yes. But today (and in your proposed solution) there's no choice.

Reply:
I agree 👍. I'm coming from web2 and in general I'm more inclined to make the API clearer and easier to use. I didn't want that to influence me much and trying to embrace optimizing for the blockchain itself as well.  @shawntabrizi  suggested using the hash as an optimization and using preimage pallet when I discussed with him during the PBA earlier. I know it's different schools but I'm interested to hear others thoughts as well.

CC @xlc @Ank4n @bkchr

Reply:
Preimage would work, but in practice it's still cumbersome because only the submitter can delete/free the preimage that was set. This means that the submitter would need an additional call, once the multisig proposal has been executed, to free their preimage. My choice would be to have the same logic as the preimage, and the same deposit, but this preimage would be linked to the multisig proposal, and would be freed when the call is executed.

Reply:
We can just have both. Make it something like

```rs
enum CallOrHash {
  Call(Vec<u8>,
  Hash(H256),
}
```

and up to the user to decide how to use it

In fact, this is how the current multisig works

Reply:
I was thinking having the `CallOrHash` will be complicating the solution more when I started working on this. Now I see it might be better to use it with some preimage in the state as well. I will change the RFC accordingly then.

Reply:
I've changed the RFC to use `CallOrHash` in all extrinsics.

>  I'd strongly advocate to include a mechanism to optionally store the call in the state, for start_proposal and then have it optional as well for the execute_proposal and retrieve it from the state if possible.

@Tbaut 
I've not written the details of storing the preimage but the `execute_proposal` specs explains it and added a new error if the preimage is not found by the time of execution.